From 4e4ad372fc17507fd32d0967987d310eafd4ef3d Mon Sep 17 00:00:00 2001
From: Thao Luong <thao.luong.uw@rvc.renesas.com>
Date: Fri, 18 Jul 2014 10:32:40 +0700
Subject: [PATCH 20/22] Add omxh264enc

Support omxh264enc for RCarGen2. omxh264enc receive video/x-raw with format is
{NV12, I420} and encode to video/x-264, stream-format is byte-stream. For more
information please refer to gst-inspect-1.0 omxh264enc
Ex:
gst-launch-1.0 -e videotestsrc ! video/x-raw,width=640,height=480,format=NV12 ! omxh264enc target-bitrate=4800000 control-rate=2 ! h264parse ! video/x-h264,stream-format=avc,alignment=au ! qtmux ! filesink location=encodefile.mp4

refs #58093
Signed-off-by: Thao Luong <thao.luong.uw@rvc.renesas.com>
---
 config/rcar/gstomx.conf |    8 ++++++++
 omx/gstomxh264enc.c     |    9 +++++++--
 omx/gstomxvideoenc.c    |   17 ++++++++---------
 3 files changed, 23 insertions(+), 11 deletions(-)

diff --git a/config/rcar/gstomx.conf b/config/rcar/gstomx.conf
index c81055a..b294e6a 100644
--- a/config/rcar/gstomx.conf
+++ b/config/rcar/gstomx.conf
@@ -16,4 +16,12 @@ in-port-index=0
 out-port-index=1
 hacks=
 
+[omxh264enc]
+type-name=GstOMXH264Enc
+core-name=/usr/local/lib/libomxr_core.so
+component-name=OMX.RENESAS.VIDEO.ENCODER.H264
+rank=256
+in-port-index=0
+out-port-index=1
+hacks=
 
diff --git a/omx/gstomxh264enc.c b/omx/gstomxh264enc.c
index 9adac04..7f78270 100644
--- a/omx/gstomxh264enc.c
+++ b/omx/gstomxh264enc.c
@@ -1,6 +1,7 @@
 /*
  * Copyright (C) 2011, Hewlett-Packard Development Company, L.P.
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>, Collabora Ltd.
+ * Copyright (C) 2014, Renesas Electronics Corporation
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -61,7 +62,11 @@ gst_omx_h264_enc_class_init (GstOMXH264EncClass * klass)
   videoenc_class->get_caps = GST_DEBUG_FUNCPTR (gst_omx_h264_enc_get_caps);
 
   videoenc_class->cdata.default_src_template_caps = "video/x-h264, "
-      "width=(int) [ 16, 4096 ], " "height=(int) [ 16, 4096 ]";
+      "width=(int) [ 16, 4096 ], " "height=(int) [ 16, 4096 ], "
+      "stream-format= (string)byte-stream, "
+      "profile = (string){baseline, main, high}, "
+      "level = (string){1, 1b, 1.1, 1.2, 1.3, 2, 2.1, 2.2, \
+                                             3, 3.1, 3.2, 4, 4.1, 4.2}";
   videoenc_class->handle_output_frame =
       GST_DEBUG_FUNCPTR (gst_omx_h264_enc_handle_output_frame);
 
@@ -69,7 +74,7 @@ gst_omx_h264_enc_class_init (GstOMXH264EncClass * klass)
       "OpenMAX H.264 Video Encoder",
       "Codec/Encoder/Video",
       "Encode H.264 video streams",
-      "Sebastian Dröge <sebastian.droege@collabora.co.uk>");
+      "Encode video with OpenMAX IL <Renesas Electronics>");
 
   gst_omx_set_default_role (&videoenc_class->cdata, "video_encoder.avc");
 }
diff --git a/omx/gstomxvideoenc.c b/omx/gstomxvideoenc.c
index d4d001e..0ae3305 100644
--- a/omx/gstomxvideoenc.c
+++ b/omx/gstomxvideoenc.c
@@ -1,6 +1,7 @@
 /*
  * Copyright (C) 2011, Hewlett-Packard Development Company, L.P.
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>, Collabora Ltd.
+ * Copyright (C) 2014, Renesas Electronics Corporation
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -194,7 +195,9 @@ gst_omx_video_enc_class_init (GstOMXVideoEncClass * klass)
 
   klass->cdata.default_sink_template_caps = "video/x-raw, "
       "width = " GST_VIDEO_SIZE_RANGE ", "
-      "height = " GST_VIDEO_SIZE_RANGE ", " "framerate = " GST_VIDEO_FPS_RANGE;
+      "height = " GST_VIDEO_SIZE_RANGE ", "
+      "framerate = " GST_VIDEO_FPS_RANGE ","
+      "format=(string) {I420, NV12}";
 
   klass->handle_output_frame =
       GST_DEBUG_FUNCPTR (gst_omx_video_enc_handle_output_frame);
@@ -1193,7 +1196,7 @@ gst_omx_video_enc_set_format (GstVideoEncoder * encoder,
         (info->width + port_def.nBufferAlignment - 1) &
         (~(port_def.nBufferAlignment - 1));
   else
-    port_def.format.video.nStride = GST_ROUND_UP_4 (info->width);       /* safe (?) default */
+    port_def.format.video.nStride = GST_ROUND_UP_32 (info->width);
 
   port_def.format.video.nFrameHeight = info->height;
   port_def.format.video.nSliceHeight = info->height;
@@ -1263,13 +1266,9 @@ gst_omx_video_enc_set_format (GstVideoEncoder * encoder,
     if (gst_omx_port_allocate_buffers (self->enc_in_port) != OMX_ErrorNone)
       return FALSE;
 
-    /* And disable output port */
-    if (gst_omx_port_set_enabled (self->enc_out_port, FALSE) != OMX_ErrorNone)
-      return FALSE;
-
-    if (gst_omx_port_wait_enabled (self->enc_out_port,
-            1 * GST_SECOND) != OMX_ErrorNone)
-      return FALSE;
+     /* Allocate for output port */
+   if (gst_omx_port_allocate_buffers (self->enc_out_port) != OMX_ErrorNone)
+       return FALSE;
 
     if (gst_omx_component_get_state (self->enc,
             GST_CLOCK_TIME_NONE) != OMX_StateIdle)
-- 
1.7.9.5

